src/Std/Data/HashMap/Lemmas.lean

```diff
@@ -1269,4 +1269,330 @@ theorem isEmpty_unitOfList [EquivBEq α] [LawfulHashable α]
+    (alter m k f).size ≤ m.size + 1 :=
+  DHashMap.Const.size_alter_le_size
+
+theorem size_le_size_alter [LawfulBEq α] {k : α} {f : Option β → Option β} :
+    m.size - 1 ≤ (alter m k f).size :=
+  DHashMap.Const.size_le_size_alter
+
+theorem get?_alter [EquivBEq α] [LawfulHashable α] {k k' : α} {f : Option β → Option β} :
+    get? (alter m k f) k' =
+      if k == k' then
+        f (get? m k)
+      else
+        get? m k' :=
+  DHashMap.Const.get?_alter
+
+@[simp]
+theorem get?_alter_self [EquivBEq α] [LawfulHashable α] {k : α} {f : Option β → Option β} :
+    get? (alter m k f) k = f (get? m k) :=
+  DHashMap.Const.get?_alter_self
+
+theorem get_alter [EquivBEq α] [LawfulHashable α] {k k' : α} {f : Option β → Option β}
+    {h : k' ∈ alter m k f} :
+    get (alter m k f) k' h =
+      if heq : k == k' then
+        haveI h' : (f (get? m k)).isSome := mem_alter_of_beq heq |>.mp h
+        f (get? m k) |>.get h'
+      else
+        haveI h' : k' ∈ m := mem_alter_of_beq_eq_false (Bool.not_eq_true _ ▸ heq) |>.mp h
+        get m k' h' :=
+  DHashMap.Const.get_alter
+
+@[simp]
+theorem get_alter_self [EquivBEq α] [LawfulHashable α] {k : α} {f : Option β → Option β}
+    {h : k ∈ alter m k f} :
+    haveI h' : (f (get? m k)).isSome := mem_alter_self |>.mp h
+    get (alter m k f) k h = (f (get? m k)).get h' :=
+  DHashMap.Const.get_alter_self
+
+theorem get!_alter [EquivBEq α] [LawfulHashable α] {k k' : α} [Inhabited β]
+    {f : Option β → Option β} : get! (alter m k f) k' =
+      if k == k' then
+        f (get? m k) |>.get!
+      else
+        get! m k' :=
+  DHashMap.Const.get!_alter
+
+@[simp]
+theorem get!_alter_self [EquivBEq α] [LawfulHashable α] {k : α} [Inhabited β]
+    {f : Option β → Option β} : get! (alter m k f) k = (f (get? m k)).get! :=
+  DHashMap.Const.get!_alter_self
+
+theorem getD_alter [EquivBEq α] [LawfulHashable α] {k k' : α} {v : β} {f : Option β → Option β} :
+    getD (alter m k f) k' v =
+      if k == k' then
+        f (get? m k) |>.getD v
+      else
+        getD m k' v :=
+  DHashMap.Const.getD_alter
+
+@[simp]
+theorem getD_alter_self [EquivBEq α] [LawfulHashable α] {k : α} {v : β} {f : Option β → Option β} :
+    getD (alter m k f) k v = (f (get? m k)).getD v :=
+  DHashMap.Const.getD_alter_self
+
+theorem getKey?_alter [EquivBEq α] [LawfulHashable α] {k k' : α} {f : Option β → Option β} :
+    (alter m k f).getKey? k' =
+      if k == k' then
+        if (f (get? m k)).isSome then some k else none
+      else
+        m.getKey? k' :=
+  DHashMap.Const.getKey?_alter
+
+theorem getKey?_alter_self [EquivBEq α] [LawfulHashable α] {k : α} {f : Option β → Option β} :
+    (alter m k f).getKey? k = if (f (get? m k)).isSome then some k else none :=
+  DHashMap.Const.getKey?_alter_self
+
+theorem getKey!_alter [EquivBEq α] [LawfulHashable α] [Inhabited α] {k k' : α}
+    {f : Option β → Option β} : (alter m k f).getKey! k' =
+      if k == k' then
+        if (f (get? m k)).isSome then k else panic ""
```

Comment:
Comparing with `Option.get!_none`, perhaps just writing `default` instead of `panic ""` is fine here.